### PR TITLE
Add proper AMD Integrated Graphics detection.

### PR DESCRIPTION
**Pull request by Zormeister**
As the title implies, it's a proper implementation for detecting the iGPU device.

This targets more "_recent_" iGPUs.

Notably those of the:
- Trinity and Richland family,
- Kaveri/Kabini/Mullins Family,
- Carrizo and Stoney family,
- Mobile Vega ASIC family

### CODE REVIEW

File: Lilu/Sources/kern_devinfo.cpp [hunk 1]
```diff
@@ -206,9 +206,17 @@ void DeviceInfo::grabDevicesFromPciRoot(IORegistryEntry *pciRoot) {
 				continue;
 
 			if (vendor == WIOKit::VendorID::Intel && (code == WIOKit::ClassCode::DisplayController || code == WIOKit::ClassCode::VGAController)) {
-				DBGLOG("dev", "found IGPU device %s", safeString(name));
-				videoBuiltin = obj;
-				requestedExternalSwitchOff |= videoBuiltin->getProperty(RequestedExternalSwitchOffName) != nullptr;
+					DBGLOG("dev", "found IGPU device %s", safeString(name));
```
Comment (vit9696) [inline Lilu/Sources/kern_devinfo.cpp:209]: Why did you change indentation for this code?

File: Lilu/Sources/kern_devinfo.cpp [hunk 2]
```diff
@@ -206,9 +206,17 @@ void DeviceInfo::grabDevicesFromPciRoot(IORegistryEntry *pciRoot) {
...
+					videoBuiltin = obj;
+					requestedExternalSwitchOff |= videoBuiltin->getProperty(RequestedExternalSwitchOffName) != nullptr;
+			} else if (vendor == WIOKit::VendorID::ATIAMD && (code == WIOKit::ClassCode::DisplayController || code == WIOKit::ClassCode::VGAController)) {
+				uint32_t dev = 0;
+				WIOKit::getOSDataValue(obj, "device-id", dev);
+				dev = (dev & 0xFF00);
+				if ((dev == 0x1600 || dev == 1500) || (dev == 0x9800 || dev == 0x1300) || (dev == 0x9900 || dev == 0x9600)) {
```
Comment (vit9696) [inline Lilu/Sources/kern_devinfo.cpp:216]: `dev == 1500` makes no sense, since `dev & 0xFF00` will strip down lower 8 bits and `1500` in hexadecimal is `0x5DC`.

Comment (vit9696) [inline Lilu/Sources/kern_devinfo.cpp:216]: In general I disagree with hard-coding device IDs in this file. Is it possible to avoid it?

**Review by vit9696 — COMMENTED** [review]
Where does the current code break that you added hardcoded device identifiers, which are going to break far more often as new hardware revisions are released?

File: Lilu/Sources/kern_devinfo.cpp (hunk 2 above)
Comment (Zormeister) [reply Lilu/Sources/kern_devinfo.cpp:216]: That’s actually my mistake, It’s meant to be hexadecimal

**Comment by Zormeister** [comment]
> Where does the current code break that you added hardcoded device identifiers, which are going to break far more often as new hardware revisions are released?

The old HDA detection code breaks for legacy systems somewhere down the line, AMD’s Device IDs and GC versions are pretty hardcoded in of themselves, you _have_ to know what versions (GC) map to a device or get the APU flag set under AMDGPU

File: Lilu/Sources/kern_devinfo.cpp (hunk 1 above)
Comment (Zormeister) [reply Lilu/Sources/kern_devinfo.cpp:209]: accidental change i believe

**Comment by vit9696** [comment]
I.e. this is entirely impossible to distinguish discrete PCI GPU from APU based on topology on AMD?

File: Lilu/Sources/kern_devinfo.cpp [hunk 3]
```diff
@@ -209,6 +209,14 @@ void DeviceInfo::grabDevicesFromPciRoot(IORegistryEntry *pciRoot) {
 				DBGLOG("dev", "found IGPU device %s", safeString(name));
 				videoBuiltin = obj;
 				requestedExternalSwitchOff |= videoBuiltin->getProperty(RequestedExternalSwitchOffName) != nullptr;
+			} else if (vendor == WIOKit::VendorID::ATIAMD && (code == WIOKit::ClassCode::DisplayController || code == WIOKit::ClassCode::VGAController)) {
+				uint32_t dev = 0;
+				WIOKit::getOSDataValue(obj, "device-id", dev);
+				dev = (dev & 0xFF00);
+				if ((dev == 0x1600 || dev == 0x1500) || (dev == 0x9800 || dev == 0x1300) || (dev == 0x9900 || dev == 0x9600)) {
```
Comment (vit9696) [inline Lilu/Sources/kern_devinfo.cpp:216]: If we end up having this code, these numbers must become named constants, so that it is clear which id belongs to which family.

**Comment by Zormeister** [comment]
> I.e. this is entirely impossible to distinguish discrete PCI GPU from APU based on topology on AMD?

AFAIK, The only way to do it on the PCI device is by doing what I did, AM4 iGPUs exist so it isn’t reliable to rely on the CPUID PackageType data

File: Lilu/Sources/kern_devinfo.cpp (hunk 3 above)
Comment (Zormeister) [reply Lilu/Sources/kern_devinfo.cpp:216]: As for this, the generic &ing makes the Device ID range more generic so there isn’t a million cases.

0x1500 is commonly associated with AMD’s Raven line, 0x1600 catches devices from Renoir, Van Gogh and Rembrandt

0x1300 catches Kaveri (Spectre) and a certain Granite Ridge iGPU (0x13C0)

0x9800 catches Carrizo, Mullins (Godavari), Kabini (Kalindi) and Stoney

0x9600 and 0x9900 catch some of the legacy TeraScale iGPUs such as Trinity and Richland iirc

Comment (vit9696) [reply Lilu/Sources/kern_devinfo.cpp:216]: This information must be in the code, not in the comments to the review. Unnamed numeric literals are plain bad, make them named constants.

Comment (Zormeister) [reply Lilu/Sources/kern_devinfo.cpp]: was going to do this already, got caught up in other things
pushed out the changes now

**Comment by Zormeister** [comment]
> i lke the idea but why not use linux constants ? this would help in future updates.
> 
> from amd_asic_type.h enum amd_asic_type { CHIP_TAHITI = 0, CHIP_PITCAIRN, /* 1 _/ CHIP_VERDE, /_ 2 _/ CHIP_OLAND, /_ 3 */ (...)
> 
> and from amdgpu_drv.c /* Navi14 */ {0x1002, 0x7340, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_NAVI14}, {0x1002, 0x7341, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_NAVI14}, {0x1002, 0x7347, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_NAVI14}, {0x1002, 0x734F, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_NAVI14},
> 
> ```
> /* Renoir */
> {0x1002, 0x15E7, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RENOIR|AMD_IS_APU},
> {0x1002, 0x1636, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RENOIR|AMD_IS_APU},
> {0x1002, 0x1638, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RENOIR|AMD_IS_APU},
> {0x1002, 0x164C, PCI_ANY_ID, PCI_ANY_ID, 0, 0, CHIP_RENOIR|AMD_IS_APU},
> ```
> 
> (...)
> 
> note that the flags also help cause we got AMD_IS_APU for proper devices

the generic constants help to reduce complexity, but also Linux isn't a reliable source since APU detection shifted to using the IP Discovery binary and marking the device as an APU from the GC version, see [here](https://github.com/ROCm/ROCK-Kernel-Driver/blob/master/drivers/gpu/drm/amd/amdgpu/amdgpu_discovery.c#L2667)

devices such as Granite Ridge and Raphael aren't in the AMDGPU driver itself, leading to whoever wanting to find the Device ID to have to use external places, such as `pcilookup.com`, `pciids`, `GPUOpen-Tools/device_info` (and even then, this specific one hasn't gotten an update for a while), etc.
